Plans: Add parameter to show only specific plans for the Jetpack app plan selection flow

staskus opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #76968

### Details

pcdRpT-2Bo-p2
Scope: pc8eDl-Vv-p2
Discussion: p1684158877461119-slack-CFFF01Q4V

Create a WP.com Plans selection for the Jetpack app limiting it to only showing WP.com Personal and Premium plans.

## Proposed Changes

Check `jetpackAppPlans` parameter to set `planTypes` props of `PlansFeaturesMain` component to limit plan selection to Personal and Premium Plans:
- /plans/{siteURL}?**jetpackAppPlans=true**
- /plans/yearly/{siteURL}?domainAndPlanPackage=true&**jetpackAppPlans=true**

Follow a similar pattern to #66022 which added `domainAndPlanPackage` parameter

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Normal plan selection flow
1. Select simple/atomic site
2. Visit /plans/{siteURL}?**jetpackAppPlans=true** page
3. Confirm that only personal and premium plans are visible
4. Switch between annual and monthly plans
5. Confirm that only personal and premium plans are visible
6. Confirm that after selecting the plan, the checkout view is opened

### Domain and plan package plan selection flow
1. Select simple/atomic site
2. Visit /plans/yearly/{siteURL}?**domainAndPlanPackage=true**&**jetpackAppPlans=true** page
3. Confirm that only personal and premium plans are visible
4. Confirm that header with steps and back button is not visible
5. Confirm that after selecting the plan, the checkout view is opened

### Regression normal plan selection flow
1. Select simple/atomic site
2. Visit /plans page
3. Confirm that all the available plans are offered

### Regression domain and plan package plan selection flow
1. Select simple/atomic site
2. Visit /plans/yearly/{siteURL}?**domainAndPlanPackage=true** page
3. Confirm that all the available plans are offered
4. Confirm that header with steps and back button appears on top

## Videos

### Before. Plans for domain and plan package

https://github.com/Automattic/wp-calypso/assets/4062343/108e0f26-593e-44e8-8961-53dcd6b53ebc

### After. Plans for domain and plan package and `jetpackAppPlans=true`

https://github.com/Automattic/wp-calypso/assets/4062343/08acb533-ddb8-4b5a-a926-1406eb068bee

### After. Plans with `jetpackAppPlans=true`

https://github.com/Automattic/wp-calypso/assets/4062343/ccc040dc-6dc3-4d2a-93ff-056e261f1635

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

staskus committed:
Show only personal and premium plans when jetpackAppPlans parameter is set

Support configuring wp.com/plans/yearly/{siteURL} flow with jetpackAppPlans parameter, that limits the number of plans shown to support Jetpack app-specific flows

staskus committed:
Hide domain and plan package header from Plans when jetpackAppPlans parameter is set

Support the Jetpack app plan purchasing flows by hiding a navigation header that displays the position inside domain and plan upsell flow and allows to come back

staskus committed:
Remember jetpackAppPlans parameter when toggling between Plan intervals

staskus commented:
@southp, we would be thankful if you could take a look at this PR when you're available. Since we're not contributing to Calypso often, let us know if anything is missing. Thank you! 

CC: @ravishanker, @irfano, @guarani

guarani reviewed (commented):
👋 This looks good @staskus! 
There might be a convention that values of `true` and `false` for "boolean" query params are equivalent, but I haven't found one. However, `jetpackAppPlans` follows the same convention set by `domainAndPlanPackage`, so this seems OK.

I tested locally using [the quick start testing steps](https://github.com/Automattic/wp-calypso/blob/trunk/docs/install.md#quick-summary-of-steps) and it worked as per the test plan. I can't see any potential regressions so I think this is good from my side.

Let me know if I should provide an approval or leave that for another reviewer.

staskus committed:
Parse domainAndPlanPackage and jetpackAppPlans parameters to support both true and false values

Parse using JSON.parse so parameter=true would be correctly parsed as a true Boolean and parameter=false would be parsed as a false Boolean

staskus committed:
Merge branch 'trunk' into update/76968-different-plans-selection-for-the-jp-app

staskus commented:
> There might be a convention that values of `true` and `false` for "boolean" query params are equivalent, but I haven't found one. However, `jetpackAppPlans` follows the same convention set by `domainAndPlanPackage`, so this seems OK.

@guarani, you spotted it well. I don't think that's correct, even if it was unharmful for `domainAndPlanPackage`. I made a fix to parse values to Booleans.

staskus committed:
Handle parameters being null before parsing to Boolean

southp commented:
@staskus Sorry for the late response. I've tested it and it works well in general. However, I'm running out of time to look into the code deeper and to go through potential regression parts. I'll ping the team for a more timely review 🙇🏼

jeyip commented:
## Testing
### Requirements

**Normal plan selection flow**
- [x] Visit /plans/{siteURL}?jetpackAppPlans=true page. Confirm that only personal and premium plans are visible.
- [x] Switch between annual and monthly plans. Confirm that only personal and premium plans are visible.
- [x] Confirm that the plans comparison grid only shows personal and premium plans.
- [x] Confirm that after selecting the plan, the checkout view is opened

**Domain and plan package plan selection flow**
- [x] Visit /plans/yearly/{siteURL}?domainAndPlanPackage=true&jetpackAppPlans=true page. Confirm that only personal and premium plans are visible.
- [ ] Confirm that header with steps and back button is not visible
Confirm that after selecting the plan, the checkout view is opened

**Regression normal plan selection flow**
Visit /plans page
Confirm that all the available plans are offered

**Regression domain and plan package plan selection flow**
Visit /plans/yearly/{siteURL}?domainAndPlanPackage=true page
Confirm that all the available plans are offered
Confirm that header with steps and back button appears on top

### Browsers
- [x] Chrome

jeyip commented on client/my-sites/plans/controller.jsx:
```diff
@@ -38,7 +38,8 @@ export function plans( context, next ) {
 					: undefined
 			}
-			domainAndPlanPackage={ context.query.domainAndPlanPackage }
+			domainAndPlanPackage={ JSON.parse( context.query.domainAndPlanPackage ?? false ) }
```
`Non-blocking / nitpick`: 

Two notes here:
1. If we want to keep JSON.parse, I think this could be more readable as `JSON.parse( Boolean( context.query.domainAndPlanPackage ) )`
2. It appears you're using JSON.parse to handle string types that express boolean values E.G. `"true"` instead of `true`. I would actually prefer `context.query.domainAndPlanPackage === "true"` because it more clearly represents our intentions of coercing some value. I'm unsure about any conventions either

staskus replied:
Thanks! That's is a good idea to explicitly check for the `"true"` case rather than doing parsing and conversion that obscures the readability of the code 👍

jeyip reviewed (approved):
I haven't working in `plans` for too long, but from what I can tell, this tests well, and I only have smaller non-blocking concerns. 🚢

jeyip commented on client/my-sites/plans/main.jsx:
```diff
@@ -259,6 +261,7 @@ class Plans extends Component {
 		}
 
 		const hideFreePlan = ! is2023PricingGridVisible || this.props.isDomainAndPlanPackageFlow;
+		const planTypes = this.props.jetpackAppPlans ? [ TYPE_PERSONAL, TYPE_PREMIUM ] : null;
```
`Non-blocking / nitpick`:

Spitballing here, but wondering if a small inline comment might be useful?

`// The jetpack mobile app only wants to display two plans -- personal and premium`

staskus replied:
Yes, I think the more explicit we are in such cases the better so the comment would only increase the clarity by expressing intentions 👍

staskus committed:
Use explicit value comparison for query parameters

staskus committed:
Add a comment explaining filtering of premium and personal plans for the Jetpack app

staskus committed:
Define domainAndPlanPackage prop as boolean

staskus committed:
Merge branch 'trunk' into update/76968-different-plans-selection-for-the-jp-app

guarani commented:
From the [Contributing docs](https://github.com/Automattic/wp-calypso/blob/trunk/docs/CONTRIBUTING.md), this needs a **[Status] Ready to Merge** label from a reviewer for it to be ready to merge.

jeyip commented:
> Is it enough to merge or are any additional actions required for the changes to be deployed?

You should be good to go!

> From the [Contributing docs](https://github.com/Automattic/wp-calypso/blob/trunk/docs/CONTRIBUTING.md), this needs a [Status] Ready to Merge label from a reviewer for it to be ready to merge.

@guarani 

I’m not sure if there’s something particular to the part of the codebase updated in this PR, but I haven’t used the  “Ready to Merge” label in any Calypso dev workflows before. Apologies if it was blocking progress on the PR, but, for me, everything seems fine to merge 🚢

guarani commented:
> I’m not sure if there’s something particular to the part of the codebase updated in this PR

I might have mis-read the docs or come across outdated ones, because I'm new to the Calypso codebase here. 

No worries and thanks for the help on this PR, @jeyip! ❤️ 

I'm going to go ahead and merge now. Nice work on this, @staskus!
